yapapi/executor/__init__.py

```diff
@@ -231,6 +232,140 @@ async def submit(
             computation_finished.set()
             self._active_computations.remove(computation_finished)
 
+    @dataclass
+    class SubmissionState:
+        """Variables related to a single call to `Executor.submit()`.
+
+        One day this will be a fully-fledged and mature class with
+        methods of its own. But for now it only includes the variables
+        required by `find_offers()`.
+        """
+
+        builder: DemandBuilder
+        agreements_pool: AgreementsPool
+        offers_collected: int = field(default=0)
+        proposals_confirmed: int = field(default=0)
+
+    async def _handle_proposal(
+        self,
+        state: "Executor.SubmissionState",
+        proposal: OfferProposal,
+    ) -> events.Event:
+        """Handle a single `OfferProposal`.
+
+        A `proposal` is scored and then can be rejected, responded with
+        a counter-proposal or stored in an agreements pool to be used
+        for negotiating an agreement.
+        """
+        score = await self._strategy.score_offer(proposal, state.agreements_pool)
+        logger.debug(
+            "Scored offer %s, provider: %s, strategy: %s, score: %f",
+            proposal.id,
+            proposal.props.get("golem.node.id.name"),
+            type(self._strategy).__name__,
+            score,
+        )
+
+        if score < SCORE_NEUTRAL:
+            reason = "Score too low"
+            await proposal.reject(reason=reason)
+            return events.ProposalRejected(prop_id=proposal.id, reason=reason)
+
+        if not proposal.is_draft:
+            # Proposal is not yet a draft of an agreement
+
+            # Check if any of the supported payment platforms matches the proposal
+            common_platforms = self._get_common_payment_platforms(proposal)
+            if common_platforms:
+                state.builder.properties["golem.com.payment.chosen-platform"] = next(
+                    iter(common_platforms)
+                )
+            else:
+                # reject proposal if there are no common payment platforms
```

Comment:
seems like there could be a `_reject_proposal(proposal, reason)` method to handle the proposal rejections...

Reply:
See 22be8cd60b878b2bbb30d431bcb9f435318c34a3